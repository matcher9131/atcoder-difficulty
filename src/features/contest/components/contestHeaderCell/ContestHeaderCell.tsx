import clsx from "clsx";
import type { ReactNode } from "react";
import { cellClassNames } from "../../../../common/cellClassNames";

export type ContestHeaderCellProps = {
    readonly textColor: string;
    readonly displayName: string;
    readonly linkHref: string;
    readonly onStatsButtonClick: () => void;
    readonly statsIconHref: string;
};

export const ContestHeaderCell = ({
    textColor,
    displayName,
    linkHref,
    onStatsButtonClick,
    statsIconHref,
}: ContestHeaderCellProps): ReactNode => {
    return (
        <td className={clsx(cellClassNames, "gap-x-1")}>
            <span className={textColor}>â—‰</span>
            <a href={linkHref} target="_blank" rel="noreferrer" className="link link-hover link-primary truncate">
                {displayName}
            </a>
            <button onClick={onStatsButtonClick} className="hidden group-hover:block group-focus-within:block">
                <svg
                    role="img"
                    aria-label="Show graph"
                    className="h-6 w-6 fill-current cursor-pointer hover:invert-30 transition-all"
                >
                    <title>Show stats</title>
                    <use href={statsIconHref} />
                </svg>
            </button>
        </td>
    );
};
