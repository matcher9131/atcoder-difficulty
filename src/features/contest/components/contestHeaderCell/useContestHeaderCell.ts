import { useAtom } from "jotai";
import type { ContestHeaderCellProps } from "./ContestHeaderCell";
import { contestMaxRatingAtom } from "../../models/contests";
import { getTextColor } from "../../functions/textColor";
import { selectedContestAtom } from "../../../contestStats/models/selectedContest";
import { useOpenModalDialog } from "../../../dialog/hooks/useOpenModalDialog";

export const useContestHeaderCell = (contestId: string): ContestHeaderCellProps => {
    const [maxRating] = useAtom(contestMaxRatingAtom(contestId));
    const textColor = getTextColor(maxRating);
    const displayName = contestId.toUpperCase();
    const linkHref = `https://atcoder.jp/contests/${contestId}`;
    const [, setSelectedContest] = useAtom(selectedContestAtom);
    const { openDialog } = useOpenModalDialog("contestStats");
    const handleStatsButtonClick = () => {
        setSelectedContest(contestId);
        openDialog().catch(console.error);
    };
    const statsIconHref = (import.meta.env.PROD ? "/atcoder-difficulty" : "") + "/resources/chart.svg#chart";
    return { textColor, displayName, linkHref, onStatsButtonClick: handleStatsButtonClick, statsIconHref };
};
