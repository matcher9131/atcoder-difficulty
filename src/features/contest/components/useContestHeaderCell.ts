import { useAtom } from "jotai";
import type { ContestHeaderCellProps } from "./ContestHeaderCell";
import { contestMaxRatingAtom } from "../dict/contests";
import { getTextColor } from "./functions";

export const useContestHeaderCell = (
    contestId: string,
): ContestHeaderCellProps => {
    const [maxRating] = useAtom(contestMaxRatingAtom(contestId));
    const textColor = getTextColor(maxRating);
    const displayName = contestId.toUpperCase();
    const linkHref = `https://atcoder.jp/contests/${contestId}`;
    return { textColor, displayName, linkHref };
};
