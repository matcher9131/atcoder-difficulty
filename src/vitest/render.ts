import { render as originalRender, type RenderOptions } from "@testing-library/react";
import userEvent, { type UserEvent } from "@testing-library/user-event";
import type { ReactNode } from "react";

export const render = (
    element: ReactNode,
    options?: Omit<RenderOptions, "queries">,
): ReturnType<typeof originalRender> & { readonly user: UserEvent } => {
    return {
        user: userEvent.setup(),
        ...originalRender(element, options),
    };
};
