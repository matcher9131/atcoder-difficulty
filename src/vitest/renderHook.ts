import { renderHook as originalRenderHook } from "@testing-library/react";
import { Provider } from "jotai";

export const renderHook = <T, Props>(
    render: (initialProps: Props) => T,
): ReturnType<typeof originalRenderHook<T, Props>> => {
    return originalRenderHook(render, { wrapper: Provider });
};
