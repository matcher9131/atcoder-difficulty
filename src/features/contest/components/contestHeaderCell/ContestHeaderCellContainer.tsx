import type { ReactNode } from "react";
import { ContestHeaderCell } from "./ContestHeaderCell";
import { useContestHeaderCell } from "./useContestHeaderCell";

type ContestHeaderCellContainerProps = {
    readonly contestId: string;
};

export const ContestHeaderCellContainer = ({
    contestId,
}: ContestHeaderCellContainerProps): ReactNode => {
    const props = useContestHeaderCell(contestId);
    return <ContestHeaderCell {...props} />;
};
