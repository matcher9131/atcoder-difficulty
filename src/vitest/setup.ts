import * as matchers from "@testing-library/jest-dom/matchers";
import type { TestingLibraryMatchers } from "@testing-library/jest-dom/matchers";
import { cleanup } from "@testing-library/react";
import { afterEach, expect } from "vitest";

// Extend Vitest's expect with jest-dom matchers
expect.extend(matchers);

// Merge type declaration of matchers of jest-dom to vitest
declare module "vitest" {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    interface Assertion<T = any> extends jest.Matchers<void, T>, TestingLibraryMatchers<T, void> {}
}

// Clean up the DOM after each test
afterEach(() => {
    cleanup();
});
