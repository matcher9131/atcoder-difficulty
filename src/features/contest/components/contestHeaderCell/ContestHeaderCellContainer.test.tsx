import { describe, it, expect, vi, beforeEach } from "vitest";
import { render } from "../../../../vitest/render";
import { ContestHeaderCellContainer } from "./ContestHeaderCellContainer";
import { useContestHeaderCell } from "./useContestHeaderCell";
import { ContestHeaderCell } from "./ContestHeaderCell";

// src/features/contest/components/contestHeaderCell/ContestHeaderCellContainer.test.tsx

// Mock ContestHeaderCell and useContestHeaderCell
const mockProps = {
    textColor: "text-blue-500",
    displayName: "Mock Contest",
    linkHref: "https://mock.com",
    onStatsButtonClick: vi.fn(),
    statsIconHref: "#icon-mock",
};

vi.mock("./useContestHeaderCell", () => ({
    useContestHeaderCell: vi.fn(),
}));

vi.mock("./ContestHeaderCell", () => ({
    ContestHeaderCell: vi.fn(() => <div data-testid="contest-header-cell" />),
}));

describe("ContestHeaderCellContainer", () => {
    const contestId = "abc123";

    beforeEach(() => {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        (useContestHeaderCell as any).mockClear();
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        (ContestHeaderCell as any).mockClear();
    });

    it("calls useContestHeaderCell with contestId", () => {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        (useContestHeaderCell as any).mockReturnValue(mockProps);
        render(<ContestHeaderCellContainer contestId={contestId} />);
        expect(useContestHeaderCell).toHaveBeenCalledWith(contestId);
    });

    it("renders ContestHeaderCell with props from hook", () => {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        (useContestHeaderCell as any).mockReturnValue(mockProps);
        render(<ContestHeaderCellContainer contestId={contestId} />);
        expect(ContestHeaderCell).toHaveBeenCalledWith(mockProps, undefined);
    });

    it("matches snapshot", () => {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        (useContestHeaderCell as any).mockReturnValue(mockProps);
        const { container } = render(<ContestHeaderCellContainer contestId={contestId} />);
        expect(container).toMatchSnapshot();
    });
});
