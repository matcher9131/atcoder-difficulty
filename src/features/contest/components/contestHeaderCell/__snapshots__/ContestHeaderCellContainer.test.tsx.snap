// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ContestHeaderCellContainer > matches snapshot 1`] = `
<div>
  <div
    data-testid="contest-header-cell"
  />
</div>
`;
